### DISCUSSION TITLE
Understanding realtime when my use case is events more than five minutes apart

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#552] [Q&A]
I'm trying to make sure I understand the right usage of realtime with my situation where I expect events to be delivered often more than five minutes apart.

If I make a realtime connection, and do not create anything new within five minutes, I see a message in the JS console indicating the connection failed. This seems in alignment with the documentation:

`The client will be disconnected automatically if it doesn't receive any events/messages for 5 minutes.`

But, I'm a little confused by this note, perhaps it is outdated?

https://github.com/pocketbase/pocketbase/issues/59

It says:

* The client `if your connected client doesn't receive any messages for 5 mins then the connection will be closed`.
* AND: `When using SSE in the browser, the reconnect is automatic (if the client is actually active).`

I'm confused because the second note says "if the client is active". Would the first point make it inactive if no events have been received? What constitutes an "inactive" connection?

If you establish a realtime connection and five minutes elapses, then you no longer have an active connection? Is that right?

I have a situation where I might have infrequent (more than five minute) events sent out. In this case, should I make a client side `setInterval` call where I re-establish the realtime connection if I don't receive an event?

**Comment by ganigeorgiev** [comment]
The 5 mins disconnect is to prevent leaked/forgotten connection if for some reason after closing the browser tab (or other client) the connection with the server was not terminated (in the future this interval may change).

You could read "_if the client is actually active_" as "_if the browser tab is still open_".

You don't need to handle manually the reconnect with `setInterval`. The SDK will do that for you after a close/disconect signal is received.

It's a little confusing, I agree. I'll think later how to update the documentation and rephrase it better.

**Reply by xrd** [reply]
Hmm. Ok. Perhaps it is because of the nginx issue (https://github.com/pocketbase/pocketbase/issues/488) I'm reading about, but after I see the failure message, then if I add a new record i never seem to get an update. I'm running this inside docker using dokku. I'm not entirely sure if I understand what the correct fix is or how to troubleshoot but I'm glad to hear my client side doesn't need this extra logic.

By the way, pocketbase is fantastic. So polished and such a delightful developer experience.

**Reply by xrd** [reply]
I think i was using an old version of the server and that was probably the issue. I'll check and update this thread when I've investigated. Thanks for the clarification on usage.

**Reply by ganigeorgiev** [reply]
I've just tested it locally (PocketBase v0.7.4, JS SDK v0.7.1) and I'm not able to reproduce it.

You could try openning your browser "Inspect tool > Network" tab and see if after the 5 min disconnect there are logged requests to `/api/realtime`; you should see at least 2 requests - 1 GET for connect and 1 POST for the subscriptions submit (ignoring the cors options/preflight)

**Comment by dinbtechit** [comment]
I am curious to know how pocket base Javascript SDK handles the 6 connection limit for SSE? 

> On my own projects, I use Service-Worker to open one connection and transmit the event as a message to other tabs. I am assuming you don't have the liberty of creating service workers in JavaScript SDK as there can be only one service worker for a domain. But I am super intrigued to know how PocketBase addresses this limitation? Cheers 🙌

**Reply by ganigeorgiev** [reply]
@dinbtechit There are 2 things that help:

1. The 6 connections SSE limit is only for HTTP1. PocketBase when started with the `--https` flag uses HTTP2 where the connection limit is a lot higher (~100).

2. Even if for some reason you have to use HTTP1 (eg. behind a restricted reverse proxy), the **SDKs will try to maintain only a single SSE connection per tab**. Or in other words, even if you call multiple times `client.realtime.subscribe(...)` in a page, internally it will start only 1 connection.

**Reply by dinbtechit** [reply]
That's smart! Thanks for the response! @ganigeorgiev

**Reply by jurb** [reply]
I hope my question is related enough to piggyback onto this @ganigeorgiev 

What actually happens in a users browser when they are subscribed to a value in the database and the value updates, when there are more than 100 realtime connections?

(I have a situation where I want more than 100 people to be subscribed with a realtime connection to a row in a database. It's a collective quiz scenario, where the group needs to move from question to question.)

**Reply by ganigeorgiev** [reply]
@jurb When a db record value is changed we search for all subscribed clients that satisfy the realtime topic and then broadcast the change as SSE event message to each client in a fan-out manner.
Note that the order of which client will receive first the message may not be the same as the order of when they subscribed because the broadcasting happens concurrently and it could depend on various factors like the client's network speed, general network latency, etc.

**Reply by ganigeorgiev** [reply]
_@jurb Also to avoid eventual misunderstanding, note that the above HTTP2 ~100 connections limit is for a **single client/browser**, meaning that you can have thousands+ realtime connections from different users at the same time just fine._

**Reply by jurb** [reply]
@ganigeorgiev oh man, I thought it was ~100 connections in total! Thank you so much for clarifying this. I got nothing to worry about. Thanks for this great project
